### PR TITLE
[FLINK-23209] Introduce HeartbeatListener.notifyTargetUnreachable

### PR DESCRIPTION
**Pull request by tillrohrmann**
With this PR Flink's HeartbeatServices listen to the result of the heartbeat rpcs. If a rpc fails with `RecipientUnreachableException`, then it will fail the heartbeat and call `HeartbeatListener.notifyTargetUnreachable`. All Flink components (`ResourceManager`, `JobMaster` and `TaskExecutor`) will treat this signal similar to a heartbeat timeout. This means that connection failures on the TCP layer will now speed up the detection of lost components.

The basic change happens in the `HeartbeatMangerImpl` and `HeartbeatManagerSenderImpl` where we listen on the result future of `HeartbeatTarget.receiveHeartbeat` and `.requestHeartbeat`. This is necessary in order to any RPC related failures to the `HeartbeatManager` so that we can react on a `RecipientUnreachableException` and trigger `HeartbeatListener.notifyTargetUnreachable`.

The PR adds tests for the main components in `JobMasterTest`, `TaskExecutorTest` and `ResourceManagerTest`. Moreover, there are tests for the `HeartbeatManager` behaviour itself in `HeartbeatManagerTest`.

This PR is based on #16342.

### CODE REVIEW

**Comment by tillrohrmann** [comment]
One big question is how we want to treat the unreachable signal. The problem is that on unstable network environments where the network connection often breaks down, listening to the unreachable signal will lead to more cancellation and re-deployments of tasks. On the other hand, it speeds up the detection of dead `TaskExecutors` significantly.

One idea could be to make it configurable and to also define a threshold for when to call `HeartbeatListener.notifyTargetUnreachable`. The main question imo is how long these intermittent networking problems exist and at what point one rather wants to restart the job than to wait on the existing TaskManager's to reconnect.

**Comment by zentol** [comment]
I'm not sure if we really want to introduce another option that does similar things as the `akka.ask.timeout`; it will make it harder to understand in which circumstances some event will cause a failure.

Currently, if a TM goes down the heartbeat mechanism needs between `heartbeatTimeout` to `heartbeatTimeout + heartbeatInterval` time to detect it; because it might go down right after a heartbeat was acknowledged.

How about we use this new mechanism to reduce this delta instead? We start the timeout clock not only when the heartbeat request is sent, but also if an RPC message cannot be delivered. Similarly, we can also stop the timeout clock if an RPC message got through, independent of what the heartbeats do.
IOW, any RPC message would be treated like a heartbeat request, and heartbeats are just a way to ensure periodic communication.

**Comment by pltbkd** [comment]
By default the heartbeat timeout is 50s while interval is 10s, so there may not be a significant improvement only to reduce the interval. 
But I find that the default `akka.ask.timeout` is 10s. Considering that the heartbeat is also sent via akka, I'm not sure whether the `heartbeat.timeout` is working.

**Comment by zentol** [comment]
> By default the heartbeat timeout is 50s while interval is 10s, so there may not be a significant improvement only to reduce the interval.

Well one needs to ask themselves why it is that the timeout is multiple times the interval.

If the timeout is that large because a target should truly only be considered unreachable if nothing got through during this entire period, then in any case both mechanism will work the same way (<= because users configure it that way).

Beyond that, it is mostly for reliability isn't it; wouldn't want to treat a TM as unreachable on the off-chance that these exact 2 messages got lost in the network. But if other messages also function as heartbeats, then a user has more leeway to reduce the timeout, increasing detection speed without compromising reliability.

> Considering that the heartbeat is also sent via akka, I'm not sure whether the heartbeat.timeout is working.

The heartbeat timeout isn't applied per-message, essentially when you send a heartbeat request you start a timer for the heartbeat timeout. If it comes back in time you stop the timer. If another heartbeat request is triggered and a timer is already running, then you don't reset it.
IOW; the heartbeat timeout describes how much time must pass during which _none_ of the heartbeats are acknowledged for the target to be considered unreachable.

**Comment by pltbkd** [comment]
> Well one needs to ask themselves why it is that the timeout is multiple times the interval.
> 
> If the timeout is that large because a target should truly only be considered unreachable if nothing got through during this entire period, then in any case both mechanism will work the same way (<= because users configure it that way).

In fact I saw that in `HeartbeatManagerOptions`, defined as defaultValue though the method is deprecated, am I at the right place?

> IOW, any RPC message would be treated like a heartbeat request, and heartbeats are just a way to ensure periodic communication.

There's a counter example that, assuming the heartbeat interval is 10s and timeout is 5s, when JM send a RPC request(not heartbeat) that costs more 5s to process in the remote TM, and there's no response of other RPC requests or heartbeat requests received during the 5s, the request will result in heartbeat timeout, though there's nothing wrong.
The main difference is that a heartbeat request is expected to be responded as soon as possible, while a RPC request may take some time before responded.

**Comment by pltbkd** [comment]
Till and I have discussed a little in the mail list on the topic, and I'd like to share a point I came up there.

> I was just considering that, since the environment is shared by JM and TMs, and the connections among TMs (using netty) are flaky in unstable environments, which will also cause the job failure, is it necessary to build a strongly guaranteed connection between JM and TMs, or it could be as flaky as the connections among TMs? As far as I know, connections among TMs will just fail on their first connection loss, so behaving like this in JM just means "as flaky as connections among TMs". 

And here's the reply from Till in the mail list.
> One simple approach could be to make the number of failed heartbeat RPCs until a target is marked as unreachable configurable because what represents a good enough criterion in one user's environment might produce too many false-positives in somebody else's environment. Or even simpler, one could say that one can disable reacting to a failed heartbeat RPC as it is currently the case.

I think it's ok to introduce a new option to determine when JM can ensure a connection loss really means a TM loss. The option is in fact a part of heartbeat configuration, rather than akka configuration, and heartbeat options are all "expert" options, indicated by the document annotation. I suppose it's not a confusing option for experts. And of course we need to provide a proper default value for normal users.

We may need another interval option if we use number of failed heartbeat RPCs as the new option. Current heartbeat interval is too long for it.

Or maybe we can use another timeout option that we can keep retrying (with a non-configurable short delay) to connect to TM until the timeout. The main difference between it and current heartbeat timeout is that, current heartbeat timeout is for liveness probe, when we suppose the TM to be living, while the new timeout is for death probe, when we suppose the TM to be already dead, and we already have some evidences about this. So it can be shorter than heartbeat timeout, which means faster failure.

**Comment by tillrohrmann** [comment]
Thanks for sharing your thoughts @zentol and @pltbkd. I think this is a very important discussion to have. In general we do agree that we should try to speed up Flink's detection of failed/dead TaskManagers. 

Currently, we only have the heartbeat mechanism for that. The current idea for the heartbeat mechanism is that the components send a heartbeat every `heartbeat.interval` ms. If one side does not receive a heartbeat for `heartbeat.timeout` ms, then it marks the other side as dead.

The way it is currently implemented is that every heartbeat signal schedules a new timeout event for in `heartbeat.timeout` ms and cancels the old one.

One idea for speeding the failure detection up is to decrease the default values for `interval` and `timeout`. Note that `interval < timeout`. However, it is currently unclear how low we can go because it strongly depends on the network environment stability of our users.

Another idea would be to use additional signals. One example could be a signal from the underlying resource manager to tell us that a `TaskExecutor` is dead (https://issues.apache.org/jira/browse/FLINK-23216). Another signal could be whether we could actually deliver our heartbeat to the target (information from the TCP stack whether there is a connection and whether the message was lost).

The connection/lost message signal is the one this PR tries to make use of. Note that we cannot solely rely on this signal because a remote `TaskExecutor` might block its main thread which makes it unresponsive but we could still deliver heartbeat requests to it (from the TCP perspective). The question is whether this signal is more predictive than others.

The connection/lost message signal can also generate false-positives because there could be a transient network split. This strongly depends on the actual user environment. Therefore, my suggestion would be to introduce a threshold for lost heartbeat messages before triggering a timeout/mark the `TaskExecutor` as dead.

Note that `TaskExecutors` will fail their tasks if there is a connection loss between `TaskExecutors`. Moreover, there is no retry for the lost connections. Hence, if the user runs Flink in an unstable environment, then she will sooner or later experience restarts because of the unstable network environment. Hence, one could argue that heartbeating could mark their targets as dead after a single lost heartbeat message because it won't produce more false-positives as Flink's data-layer will already produce.

Of course, we can also think about more complex models where all different signals are integrated to produce a confidence value and then for some value we say that the target is now "dead". Also we can think about different retry methods (e.g. changing the request interval based on the signal), but I think that simply having a threshold for lost heartbeat messages could already improve the situation for some expert users.

**Comment by zentol** [comment]
> Therefore, my suggestion would be to introduce a threshold for lost heartbeat messages before triggering a timeout/mark the TaskExecutor as dead.

I'm curious as to how this threshold would be implemented considering that there is no reliable pattern for how many/often RPC messages are sent.
It seems like any approach where X failed messages in Y time would behave undesirable in the extreme cases, such as 1 RPC per Y, or hundreds of RPCs within Y/1000. I guess it would have to be something like "If all (but at least LOWER_THRESHOLD) messages in Y time could not be delivered then the target is considered unreachable."

> Of course, we can also think about more complex models where all different signals are integrated to produce a confidence value and then for some value we say that the target is now "dead".

I think that would become too complex and difficult to understand to the user. (buy maybe I imagine something overly complicated)
However, I do like the idea of combining the information of both heartbeats and other mechanisms; something along the lines of "if 1 heartbeat was not answered within akka.ask.timeout and no RPC (with at least LOWER_THRESHOLD attempts) went through within the timeframe (heartbeat_start -> heartbeat_timeout), then we consider the target unreachable."; essentially we shortcut the heartbeat timeout, instead of having an entirely separate mechanism.

> Hence, one could argue that heartbeating could mark their targets as dead after a single lost heartbeat message because it won't produce more false-positives as Flink's data-layer will already produce.

That only applies if there are actual data connections between task executors; it does not apply to jobs without shuffles, jobs that are currently being deployed (potentially crashing the TM in the process), nor cases where the JM/all TMs go down.

**Comment by pltbkd** [comment]
I prefer to keep the mechanism simple and easy to understand, though I also agree that some expert configurations may be needed in specific scenarios. In the first version, a simple mechanism is enough for most scenarios, and current behavior can be reserved as default or alternative.

>I guess it would have to be something like "If all (but at least LOWER_THRESHOLD) messages in Y time could not be delivered then the target is considered unreachable."

I think this is a simple and clear rule, with a proper LOWER_THRESHOLD default value which users possibly don't need to modify. Time is the main concern and more environment-related, which is also approved by Liu in the mail list.

>That only applies if there are actual data connections between task executors; it does not apply to jobs without shuffles, jobs that are currently being deployed (potentially crashing the TM in the process), nor cases where the JM/all TMs go down.

It indeed increases the instability if we make connections between JM and TM as weak as that between TMs, especially when no connection exists between TMs. But I suppose it's acceptable to have the same experience in a certain cluster with different type of jobs, no matter whether data connections exist.
Moreover, in a job without data connections, tasks in other TMs won't be blocked when one TM is lost, and the failover cost is lower. So faster detecting of TM lost is not as necessary as those jobs with data connections. Maybe we don't need to worry a lot for such jobs when we build the new mechanism since current one is possibly enough.


By the way, similar to k8s' liveness probe, maybe it's better to respond heartbeat requests asap rather than queue it with other RPC requests, which I prefer to use as readiness probe. In this way, responding of heartbeat requests in a certain time can be more predictable, therefore the result of timeout is more reliable, and timeout can possibly be decreased. Though the connection loss is still not a reliable signal. 
But since I'm not experienced in akka, I can't tell if it's possible or easy to implement.

**Comment by pltbkd** [comment]
>The idea would be to establish a reconnection mechanism between the TMs. This hasn't been done yet, though. The non-trivial part is probably that we need to introduce an ack-protocol and keep a window of sent events on the sender side in order to resend events in case of lost packages.

@tillrohrmann 
This is quoted from your reply in the mail list. I'd like to make sure if this is a long term plan or is going to take actions right now. If it's in recent plan, the heartbeat mechanism may also need to be more reliable.

**Comment by tillrohrmann** [comment]
I think in the first version the threshold would solely be defined for the heartbeat rpcs (e.g. if the `HeartbeatManager` fails to send `x` heartbeat rpcs, then mark the target as dead). This entails that the time until marking someone as dead is `threshold * heartbeat interval`. I hope that this definition is easy enough to understand for our users.

>That only applies if there are actual data connections between task executors; it does not apply to jobs without shuffles, jobs that are currently being deployed (potentially crashing the TM in the process), nor cases where the JM/all TMs go down.

Agreed that in some cases this might make the cluster a tad bit less stable but given that Flink is able to run jobs with shuffles, it seems that Flink runs often enough in a stable-enough environment these days. Hence, I would suggest to go with a default of `1` heartbeat rpc loss until marking a target as dead. If users start to complain, we can still disable it.

I would leave more complex models for future improvements.

>By the way, similar to k8s' liveness probe, maybe it's better to respond heartbeat requests asap rather than queue it with other RPC requests, which I prefer to use as readiness probe. In this way, responding of heartbeat requests in a certain time can be more predictable, therefore the result of timeout is more reliable, and timeout can possibly be decreased. Though the connection loss is still not a reliable signal. But since I'm not experienced in akka, I can't tell if it's possible or easy to implement.

The problem is that the heartbeat is used to transport status information about the components. Since we require a certain order of rpcs, we cannot send the heartbeat signals easily from outside the main thread because it could lead to race conditions and outdated status information. One could try to use some logical clocks to synchronize the messages again, but this hasn't been tried yet.

> This is quoted from your reply in the mail list. I'd like to make sure if this is a long term plan or is going to take actions right now. If it's in recent plan, the heartbeat mechanism may also need to be more reliable.

There are no immediate plans for adding this feature but it will eventually be required.

So long story short, I would be in favour of adding a threshold for lost heartbeat rpcs until a target is marked as dead. The default should be `1` and it is configurable by the user if it is too aggressive on their environment. If it turns out that too many users are running Flink on unstable environments, then we might have to increase this value in the future. What do you think?

**Comment by zentol** [comment]
> This entails that the time until marking someone as dead is threshold * heartbeat interval. I hope that this definition is easy enough to understand for our users.

yes, I think that is easy to understand. Restricting it to heartbeat RPCs also ties the whole story together; you have heartbeats as the _one_ mechanism to detect dead machines, with the timeouts handling unresponsive TMs (i.e., those blocked by some long-running operation or running exceedingly slow), while the threshold handles TMs that are already dead.

> Hence, I would suggest to go with a default of 1 heartbeat rpc loss until marking a target as dead.

We can try that out, but I am apprehensive about the default of 1 (purely because expecting every message to go through is like _the_ thing everyone drills you not to do).  

> The problem is that the heartbeat is used to transport status information about the components. Since we require a certain order of rpcs, we cannot send the heartbeat signals easily from outside the main thread because it could lead to race conditions and outdated status information. One could try to use some logical clocks to synchronize the messages again, but this hasn't been tried yet.

We could also think of decoupling heartbeats and periodic status updates again. It would result in more RPCs, but the total amount of data may become less (depending on how low the previous heartbeat interval was).

**Review by zentol — COMMENTED** [review]
The changes so far look fine to me, but the threshold isn't configurable yet.

**Comment by pltbkd** [comment]
>This entails that the time until marking someone as dead is threshold * heartbeat interval. I hope that this definition is easy enough to understand for our users.

I overall agree it is a good enough plan for the first version. Maybe we should suggest user not to set the threshold to more than timeout/interval, which by default is 50s/10s=5, a relative small value. We could add the suggestion in the configuration description, if you think it's necessary as well.

>We can try that out, but I am apprehensive about the default of 1 (purely because expecting every message to go through is like the thing everyone drills you not to do).

I agree with zentol and I'd like to suggest the default value to be 2, standing for discovering and confirming.

>The problem is that the heartbeat is used to transport status information about the components. Since we require a certain order of rpcs, we cannot send the heartbeat signals easily from outside the main thread because it could lead to race conditions and outdated status information. One could try to use some logical clocks to synchronize the messages again, but this hasn't been tried yet.

I now understand why the heartbeats also need to be in order. IMO, there are some advantages to have a more reliable and frequent heartbeat mechanism, with which I think both interval and timeout could be decreased. But since I don't know many details here, I would leave this issue to experts.

**Comment by tillrohrmann** [comment]
Alright, then let's go with a default of 2. I will update this PR to include the threshold.

Maybe as a potential follow up we could think about decreasing the heartbeat timeout to `15s` and the interval to `3s`.

**Comment by tillrohrmann** [comment]
I've added another commit that adds the threshold functionality. Please take a look @zentol and @pltbkd.

File: flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatMonitorImpl.java [hunk 1]
```diff
@@ -70,6 +81,11 @@
                 "The heartbeat timeout interval has to be larger than 0.");
         this.heartbeatTimeoutIntervalMs = heartbeatTimeoutIntervalMs;
 
+        Preconditions.checkArgument(
+                failedRpcRequestsUntilUnreachable > 0 || failedRpcRequestsUntilUnreachable == -1,
+                "The number of failed heartbeat RPC requests has to be larger than 0 or -1 (deactivated).");
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatMonitorImpl.java:86]: Shouldn't we check this in the HeartbeatServices when we actually access the configuration?

File: flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatMonitorImpl.java [hunk 2]
```diff
@@ -91,13 +107,33 @@ public long getLastHeartbeat() {
...
-            heartbeatListener.notifyTargetUnreachable(resourceID);
+    public void reportHeartbeatRpcFailure() {
+        final int failedRpcRequestsSinceLastSuccess =
+                numberFailedRpcRequestsSinceLastSuccess.incrementAndGet();
+
+        if (isHeartbeatRpcFailureDetectionEnabled()
+                && failedRpcRequestsSinceLastSuccess >= failedRpcRequestsUntilUnreachable) {
+            if (state.compareAndSet(State.RUNNING, State.UNREACHABLE)) {
+                LOG.debug(
+                        "Mark heartbeat target {} as unreachable because {} heartbeat RPCs have failed.",
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatMonitorImpl.java:118]: "...have failed in a row."?

File: flink-core/src/main/java/org/apache/flink/configuration/HeartbeatManagerOptions.java [hunk 3]
```diff
@@ -32,15 +34,40 @@
...
+
+    @Documentation.Section(Documentation.Sections.EXPERT_FAULT_TOLERANCE)
+    public static final ConfigOption<Integer> HEARTBEAT_RPC_FAILURE_THRESHOLD =
+            key(HEARTBEAT_RPC_FAILURE_THRESHOLD_KEY)
+                    .intType()
+                    .defaultValue(2)
+                    .withDescription(
+                            Description.builder()
+                                    .text(
+                                            "The number of failed heartbeat RPCs until a heartbeat target is marked as unreachable. "
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/configuration/HeartbeatManagerOptions.java:59]: I think we should mention that N RPCs have to fail in a row for this to trigger.

**Review by zentol — APPROVED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatMonitorImpl.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatMonitorImpl.java:86]: I think this makes sense, yes. I'll add it there as well.

File: flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatMonitorImpl.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatMonitorImpl.java:118]: Good point. Will update it.

File: flink-core/src/main/java/org/apache/flink/configuration/HeartbeatManagerOptions.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-core/src/main/java/org/apache/flink/configuration/HeartbeatManagerOptions.java:59]: Yes, this is a good point. I will update the description.

**Comment by tillrohrmann** [comment]
Thanks a lot for the review @zentol. I've addressed your comments.
